message-transport/pkg/transport/metrics/kafka_over_kafka/producer_decorator.go

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2018-Present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka_over_kafka
+
+import (
+	"github.com/projectriff/riff/message-transport/pkg/transport"
+	"github.com/projectriff/riff/message-transport/pkg/transport/kafka"
+	"github.com/projectriff/riff/message-transport/pkg/transport/metrics"
+)
+
+const DefaultMetricsTopic = "io.projectriff.message-transport.metrics"
+
+func NewMetricsEmittingProducer(brokerAddrs []string, producerId string, metricsTopic ... string) (transport.Producer, error) {
+	delegate, err := kafka.NewProducer(brokerAddrs)
+	if err != nil {
+		return nil, err
+	}
+	metricsProducer, err := kafka.NewProducer(brokerAddrs)
+	if err != nil {
+		return nil, err
+	}
+	return metrics.NewProducer(delegate, producerId, getMetricsTopic(metricsTopic...), metricsProducer), nil
+}
+
+func getMetricsTopic(metricsTopic ...string) string {
```

Comment:
I'm curious about why this approach was taken instead of having a single `metricTopic string` argument to `NewMetricsEmittingProducer`.

Reply:
I would like the various users of the decorated interfaces to _agree_ on the name of the metrics topic, so providing a sane default avoids the users having to (a) specify a value and (b) share a constant so they all use the same value.

Reply:
I think I'm still missing something. Could a nil-check or empty-string check achieve the same end?

Reply:
No. `metricsTopic ...string` is declaring a slice. Anyway, I've added some docs to the NewMetricsEmittingProducer/Consumer constructor functions which should help demystify the vararg.